### DISCUSSION TITLE
Report - (4th Sem.) Week 1 📢

### DISCUSSION THREAD

**Discussion by giomskii** [SE-TINF22B6/DHBWhub#469] [Weekly Report]
Welcome to the ninth report of our progress! We are excited to share the latest news and developments since the start of the new
semester with you! 🥳

#### We added three new entities to our database!
- Friendship:
  Enables the establishment of friendships within the communities and allows the user to see posts from friends upfront. The table contains the attributes "requester_id" (user that sent a friendship request),  "receiver_id" (user that received a friendship) and the flag "accepted" which indicates whether the receiver has already accepted the friendship request.
- SavedPost:
  Contains the posts which have been bookmarked/saved by each user. Thus, it contains the "client_user_id" and the respective "post_id" as table attributes 
- Events:
  We have decided to only show global events within the calendar. These are saved in the "events" table. Each event has a title/"name" and a date, on which it will occur.

In the following diagram, you can see the new entities and their relations/dependencies to existing tables:
>![new-tables](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/diagrams/blog_newEntites.drawio.png?raw=true)

#### Trigger and Views
In order to reduce response delays from client-side requests, we decided to use trigger in our postgresql database which prepare and store frequently requested responses already in the background. Thus, the backend doesn't need to filter through several data or apply operations on the database, instead, it only needs to retrieve the already prepared data from the views in the database. For instance, we created the trigger "trigger_update_post_views" that executes an Select statement on the post table (e.g. retrieve only the posts from the faculty "Technik") after each CRUD operation and stores it in an external view. Our SpringBoot application is then able to retrieve the data directly from this particular view. This process is visualized in the following graph:
>![trigger-view](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/diagrams/sources/blog_TriggerAndViews.drawio.png?raw=true)

#### New and improved mockups
We have overhauled the general design of our homepage and created new / improved existing mockups e.g. by adding a new header and footer as you can see in the image below: 
>![General Design](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/c0903617-2c51-4851-8b67-448ca0397525)
  
The mockup for the login component has also been improved:
>![Login](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/684c00e9-2849-408d-b436-1e50d0244a5d)

We also created an initial mockup draft for the search bar function which you can see in the next picture:
>![Search - PC View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/e6e613e4-f316-4dda-b2a1-2fc2b8fbf808)

#### Current state of the frontend implementation
We already finished some of the changes we did to the mockups in the actual frontend implementation. This includes the overall design of the homepage, the improvement of the header as well as a functioning sorting-component, which allows the user to sort the posts by date, followings and amount of likes. You can see the change of order in the following images:
>![Filter-Component](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/d1f68674-6077-4065-8976-4b37c0ff1581)
>![Filter-Component](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/2e27d1a9-f6ba-4e7b-8f79-019b8e15583b)
    
We even implemented our own custom 'error 404'-page, on the website it is an animated graphic.
![image](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/9ac430fa-be4e-4459-a4a4-1828187fdc63)
 
 
### Architecture choice? 🤔 
Since we are using SpringBoot as an intermediate layer between the actual database and the website, it is obvious that we are moving within the boundaries of a 'REST'-architecture. Nonetheless, we aim to reach the general guiding constraints of 'REST' such as statelessness, caching or the seperation of client and server concerns. 

### Main goals and priorities in the next weeks 📈  
There are some goals we'd like to achieve in the next weeks. For once, we'd like to connect the bridge between the backend and the frontend so that we can actually use and operate with the data that lies in the database. This also includes the complete implementation of the login/registration process, as well as the first interactions between users on our forum through posts and comments. It is also a high priority to finish/create all the mockups for the remaining components so that we have a better image in our head when trying to implement new features regarding these components.

### Sprint 6: What have we achieved ❓ What's to come❔
✅ Setup Github Pages
✅ Frontend: Implemented 'Sort options'-components
✅ Frontend: Added 'page not found'-page
✅ Frontend: Refactored general design (+ mockups)
✅ Frontend: Improved header (mockup)
✅ Frontend: Created mockup for footer and search
✅ Backend: Improvement of authentification  
✅ Backend: Revising existing JUnit tests
✅ Database: Creating Trigger functions and external views
🚧 Finish use case documentation
🚧 Revise existing models and create new proposals
🚧 Finish backend authentification
🚧 Finish use case documentation (adding last mockups)
🚧 Update database with new samples
🚧 Adapt mockup for login/signup dialog
🚧 Password encryption in the backend
🚧 Remove Herobrine

Until next week 💯

Your DHBWhub Team 🚀

**Comment by LeonmAndreesen** [comment]
Good day dear team, 

an outstanding review. You are doing really well. I like your very detailed post. You have an amazing timeline and some really good plan.

**Comment by BenjaminB109** [comment]
Dear DHBWhub Team,
I think your homepage design is very good. It is clear and well structured. I believe the REST architecture is a good choice for your project.
